### PR TITLE
Initial implementation of isAlmostEqual(to:) and isAlmostZero()

### PR DESCRIPTION
**Pull request by stephentyrone**
Pending Swift-evolution review.

Pitch thread here: https://forums.swift.org/t/approximate-equality-for-floating-point/22420

### CODE REVIEW

**Comment by stephentyrone** [comment]
@swift-ci please test

**Comment by stephentyrone** [comment]
@swift-ci please test source compatibility

File: stdlib/public/core/FloatingPoint.swift [hunk 1]
```diff
@@ -2020,6 +2020,140 @@ extension FloatingPoint {
...
+  /// their bits but are not meaningfully almost equal, the computation
+  /// is likely ill-conditioned and should be reformulated.
+  ///
+  /// For more complete guidance on an appropriate choice of tolerance,
+  /// consult with a friendly numerical analyst.
+  ///
+  /// - Parameters:
+  ///   - other: the value to compare with `self`
+  ///   - tolerance: the relative tolerance to use for the comparison.
+  ///     Should be in the range [.ulpOfOne, 1).
```
Comment (nh7a) [inline stdlib/public/core/FloatingPoint.swift]:
```suggestion
  ///     Should be in the range [.ulpOfOne, 1].
```

File: stdlib/public/core/FloatingPoint.swift [hunk 2]
```diff
@@ -2020,6 +2020,140 @@ extension FloatingPoint {
...
+  ///
+  /// - Returns: `true` if `self` is almost equal to `other`; otherwise
+  ///   `false`.
+  @inlinable
+  public func isAlmostEqual(
+    to other: Self,
+    tolerance: Self = Self.ulpOfOne.squareRoot()
+    ) -> Bool {
+    // Tolerances outside of [.ulpOfOne, 1) yield well-defined but useless
```
Comment (nh7a) [inline stdlib/public/core/FloatingPoint.swift]:
```suggestion
    // Tolerances outside of [.ulpOfOne, 1] yield well-defined but useless
```

File: stdlib/public/core/FloatingPoint.swift [hunk 3]
```diff
@@ -2020,6 +2020,140 @@ extension FloatingPoint {
...
+    // results, so this is enforced by an assert rathern than a precondition.
+    assert(tolerance >= .ulpOfOne && tolerance < 1, "tolerance should be in [.ulpOfOne, 1).")
```
Comment (nh7a) [inline stdlib/public/core/FloatingPoint.swift]:
```suggestion
    assert(tolerance >= .ulpOfOne && tolerance < 1, "tolerance should be in [.ulpOfOne, 1].")
```

File: stdlib/public/core/FloatingPoint.swift (hunk 1 above)
Comment (stephentyrone) [reply stdlib/public/core/FloatingPoint.swift]: No. 1 is not a valid value. [a, b) is mathematical notation for the set of values >= a and < b.

File: stdlib/public/core/FloatingPoint.swift (hunk 3 above)
Comment (stephentyrone) [reply stdlib/public/core/FloatingPoint.swift]: As above.

File: stdlib/public/core/FloatingPoint.swift (hunk 1 above)
Comment (nh7a) [reply stdlib/public/core/FloatingPoint.swift]: Ah, sorry, didn't know that. Never mind :)

**Comment by futurejones** [comment]
Can one of the admins verify this patch?

**Pull request closed without merging**